massa-bootstrap/src/client.rs

```diff
@@ -437,15 +464,35 @@ pub async fn get_state(
         }
         return Ok(GlobalBootstrapState::new(final_state));
     }
+
+    // we filter the bootstrap list to keep only the ip addresses we are compatible with
+    let mut filtered_bootstrap_list = filter_bootstrap_list(
+        bootstrap_config.bootstrap_list.clone(),
+        bootstrap_config.bootstrap_protocol,
+    );
+
     // we are after genesis => bootstrap
     massa_trace!("bootstrap.lib.get_state.init_from_others", {});
-    if bootstrap_config.bootstrap_list.is_empty() {
+    if filtered_bootstrap_list.is_empty() {
         return Err(BootstrapError::GeneralError(
             "no bootstrap nodes found in list".into(),
         ));
     }
-    let mut shuffled_list = bootstrap_config.bootstrap_list.clone();
+
+    filtered_bootstrap_list.shuffle(&mut StdRng::from_entropy());
```

Comment:
You don't need to shuffle it now because you already shuffle it at line 494

Reply:
I decided to shuffle two times because:

- If I only shuffle after removing duplicates, then the IP protocol used for each bootstrap server will always be the same (same order as the bootstrap list), even if we should bootstrap on "Both" protocols.
- If I only shuffle before, then the order depends on the HashMap iterator (which is arbitrary and maybe not truly random).

Do you see another way to do so? Or maybe having the same protocol is not so bad?

Reply:
If I follow you here, if you shuffle it before the dedup, it will remove sometimes ipv4 and sometimes ipv6 as it's not in the order of the bootstrap list anymore ? Or there is something I miss ?

Reply:
Exactly! I think it's better to keep sometimes Ipv4 and sometimes Ipv6 if the bootstrap server exposes both.

Reply:
Ok then I think you could something like that to dedup : https://stackoverflow.com/a/47648303/10419533 that will avoid you the re-ordering of the hashmap and then the second shuffle and should optimize this.

Reply:
Good idea, done in latest commit.